chore: move warehouse handle to router

achettyiitr opened the pull request:
# Description

- Rename HandleT to Router and move in its self-defined file. 
- Also, renamed `identifies.go` and `scheduling.go ` into their `router_identifies.go` and `router_scheduling.go` router scope.
- Removed global configurations for Router.
- Added test cases for Router [84% coverage]
- Rewrote test for router_scheduling and added some more test cases.

## Notion Ticket

https://www.notion.so/rudderstacks/Move-warehouseHandleT-to-rotuer-354c3e45fc674c3e988d1a79d244a5ff?pvs=4

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

achettyiitr commented:
https://github.com/rudderlabs/rudder-server/pull/3661/files#

fracasula commented:
@achettyiitr I'm keen to review this one but we should rebase now that the original branch has been merged.

fracasula commented on warehouse/router_identities.go:
```diff
@@ -22,15 +21,13 @@ import (
 
-func Init2() {
-	config.RegisterBoolConfigVariable(false, &shouldPopulateHistoricIdentities, false, "Warehouse.populateHistoricIdentities")
+func init() {
```
Nice. Maybe since we're making changes here we can take advantage of that and simplify some of the code that is using these maps. For example we could store a `struct{}` instead of a `bool` value and we could have functions like `isDestHistoricIdentitiesPopulated` rewritten as such:

```go
func isDestHistoricIdentitiesPopulated(warehouse model.Warehouse) bool {
	populatedHistoricIdentitiesMapLock.RLock()
	_, ok := populatedHistoricIdentitiesMap[uniqueWarehouseNamespaceString(warehouse)]
	populatedHistoricIdentitiesMapLock.RUnlock()
	return ok
}
```

And:

```go
func setDestHistoricIdentitiesPopulated(warehouse model.Warehouse) {
	populatedHistoricIdentitiesMapLock.Lock()
	populatedHistoricIdentitiesMap[uniqueWarehouseNamespaceString(warehouse)] = struct{}{}
	populatedHistoricIdentitiesMapLock.Unlock()
}

func setDestHistoricIdentitiesPopulateInProgress(warehouse model.Warehouse, starting bool) {
	populatingHistoricIdentitiesProgressMapLock.Lock()
	defer populatingHistoricIdentitiesProgressMapLock.Unlock()
	if starting {
		populatingHistoricIdentitiesProgressMap[uniqueWarehouseNamespaceString(warehouse)] = struct{}{}
	} else {
		delete(populatingHistoricIdentitiesProgressMap, uniqueWarehouseNamespaceString(warehouse))
	}
}
```

And so on...

fracasula commented on warehouse/router_scheduling.go:
```diff
@@ -7,108 +7,69 @@ import (
-	startUploadAlways       bool
 	scheduledTimesCacheLock sync.RWMutex
+
+	startUploadAlways bool
```
There might be a race here :thinking: When is the `Admin.TriggerUpload()` fired? Can it run concurrently with `Router.canCreateUpload()`?

fracasula replied:
Right, it shouldn't be much of a problem then.

fracasula replied:
I'd prefer correctness over a potential benign data race ([if such things exist](https://bartoszmilewski.com/2020/08/11/benign-data-races-considered-harmful/)). Converting this to an `atomic.Bool` should be a no :brain:

achettyiitr replied:
Good point.

fracasula commented on warehouse/router_identities.go:
```diff
@@ -152,16 +149,16 @@ func (wh *HandleT) hasLocalIdentityData(warehouse model.Warehouse) (exists bool)
+	err := r.dbHandle.QueryRow(sqlStatement).Scan(&exists)
 	if err != nil {
 		// TODO: Handle this
 		panic(fmt.Errorf("Query: %s\nfailed with Error : %w", sqlStatement, err))
```
For my information, how come we `panic` here but we are handling the other errors?

fracasula replied:
A follow up task can also be created but first it'd be nice to understand the scope of this.

achettyiitr commented on warehouse/warehouse_test.go:
```diff
@@ -72,121 +54,3 @@ func getMockStats(g GinkgoTInterface) (*mock_stats.MockStats, *mock_stats.MockMe
 	return mockStats, mockMeasurement
 }
-
-func TestUploadJob_ProcessingStats(t *testing.T) {
```
This is moved to Router Processor Stats test.

achettyiitr commented on warehouse/scheduling_test.go:
```diff
@@ -1,136 +0,0 @@
-package warehouse_test
```
This is moved to router_scheduling_test.go

fracasula commented on warehouse/internal/repo/upload_test.go:
```diff
@@ -1000,3 +1000,62 @@ func TestUploads_PendingTableUploads(t *testing.T) {
+	repoUpload := repo.NewUploads(db, repo.WithNow(func() time.Time {
+		return now
+	}))
+	repoStaging := repo.NewStagingFiles(db, repo.WithNow(func() time.Time {
```
We could move this into the `success` test.

fracasula commented on warehouse/router_identities.go:
```diff
@@ -152,16 +149,16 @@ func (wh *HandleT) hasLocalIdentityData(warehouse model.Warehouse) (exists bool)
-func (wh *HandleT) hasWarehouseData(ctx context.Context, warehouse model.Warehouse) (bool, error) {
-	whManager, err := manager.New(wh.destType, wh.conf, wh.Logger, wh.stats)
+func (r *Router) hasWarehouseData(ctx context.Context, warehouse model.Warehouse) (bool, error) {
+	whManager, err := manager.New(r.destType, r.conf, r.logger, r.stats)
```
We might not need to create a new instance each time.

fracasula replied:
This is used in other places too so it might make sense to create it only once (see line 399).

fracasula commented on warehouse/router_identities.go:
```diff
@@ -152,16 +149,16 @@ func (wh *HandleT) hasLocalIdentityData(warehouse model.Warehouse) (exists bool)
+func (r *Router) hasWarehouseData(ctx context.Context, warehouse model.Warehouse) (bool, error) {
+	whManager, err := manager.New(r.destType, r.conf, r.logger, r.stats)
 	if err != nil {
 		panic(err)
```
Why not return an error here?

fracasula commented on warehouse/router_identities.go:
```diff
@@ -173,10 +170,10 @@ func (wh *HandleT) hasWarehouseData(ctx context.Context, warehouse model.Warehou
 }
 
-func (wh *HandleT) setupIdentityTables(ctx context.Context, warehouse model.Warehouse) {
+func (r *Router) setupIdentityTables(ctx context.Context, warehouse model.Warehouse) {
```
Just asking btw, not saying we have to address this in this PR but let's discuss if needed.

fracasula commented on warehouse/router_identities.go:
```diff
@@ -342,76 +339,76 @@ func (wh *HandleT) initPrePopulateDestIdentitiesUpload(warehouse model.Warehouse
-			pkgLogger.Infof("[WH]: Did not find local identity tables..")
-			pkgLogger.Infof("[WH]: Generating identity tables based on data in warehouse %s:%s", wh.destType, warehouse.Destination.ID)
-			upload = wh.initPrePopulateDestIdentitiesUpload(warehouse)
+			r.logger.Infof("[WH]: Did not find local identity tables..")
```
```suggestion
```

fracasula commented on warehouse/router_identities.go:
```diff
@@ -441,15 +438,15 @@ func (wh *HandleT) populateHistoricIdentities(ctx context.Context, warehouse mod
 		if err != nil {
-			pkgLogger.Errorf(`[WH]: Failed fetching schema from warehouse: %v`, err)
+			r.logger.Errorf(`[WH]: Failed fetching schema from warehouse: %v`, err)
 			job.setUploadError(err, model.Aborted)
```
How come we are ignoring this error? We're doing the same with other calls to `setUploadError` or `setUploadStatus`.

fracasula reviewed (commented):
Still reviewing, posting a few questions in the meantime.

fracasula commented on warehouse/router_tracker.go:
```diff
@@ -64,33 +64,33 @@ func (wh *HandleT) Track(ctx context.Context, warehouse *model.Warehouse, config
 	}
 	statKey := "warehouse_track_upload_missing"
-	wh.stats.NewTaggedStat(statKey, stats.GaugeType, tags).Gauge(0)
+	r.statsFactory.NewTaggedStat(statKey, stats.GaugeType, tags).Gauge(0)
```
Can't we keep this stat to be used later `if !exists`?

fracasula commented on warehouse/router.go:
```diff
@@ -0,0 +1,705 @@
+			}
+		}
+
+		for _, uploadJob := range uploadJobsToProcess {
+			r.setDestInProgress(uploadJob.warehouse, uploadJob.upload.ID)
+		}
+
+		for _, uploadJob := range uploadJobsToProcess {
```
Is the double loop needed here? Could we potentially set the destination as "in progress" and give the job to a worker in the same iteration?

achettyiitr commented:
@fracasula Added a [task](https://linear.app/rudderstack/issue/PIPE-118/address-review-comments-for-warehouse-identifies) for addressing the review comments for router_identifies.
Once this is merged, will make it in a separate PR.

lvrach commented on warehouse/router_identities.go:
I have noticed we are using `fmt.Sprintf` for providing argument into to the SQL query, we should use prepared statements.

fracasula reviewed (approved):
We can have a separate task for the prepared statements.
